python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+        except AttributeError:
+            raise RuntimeError("Failed to convert image to float")
+        return floatImage
+
+
+def _crossCorrelateSimulate(im, im2, n=5, border=10, sigma=5):
+    """Perform a simple xcorr from two images.
+
+    This sim code is used to estimate the bias correction used above.
+
+    xxx need to work out why this performs differently,
+    which version is correct, and whether this function needs to exist at all.
+    I don't think it does,
+    as the real code is now modular and doesn't have to run ISR.
+
+    It contains many elements of the actual code
+    above (without individual amps and ISR removal )
+    It takes two images, im and im2;
+    n the max lag of the correlation function; border, the number of border
+    pixels to discard; and sigma the sigma to use in the mean clip.
+    """
+    sctrl = afwMath.StatisticsControl()
+    sctrl.setNumSigmaClip(sigma)
+    # im = self._convertImagelikeToFloatImage(im)
+    # im2 = self._convertImagelikeToFloatImage(im2)
+
+    means1 = [0, 0]
+    means1[0] = afwMath.makeStatistics(im[border:-border, border:-border, afwImage.LOCAL],
+                                       afwMath.MEANCLIP, sctrl).getValue()
+    means1[1] = afwMath.makeStatistics(im2[border:-border, border:-border, afwImage.LOCAL],
+                                       afwMath.MEANCLIP, sctrl).getValue()
+    im -= means1[0]
+    im2 -= means1[1]
+    diff = im2.clone()
+    diff -= im.clone()
+    diff = diff[border:-border, border:-border, afwImage.LOCAL]
+    binsize = 128  # this needs to change somehow. Where should this be got from?
+    nx = diff.getWidth()//binsize
+    ny = diff.getHeight()//binsize
+    bctrl = afwMath.BackgroundControl(nx, ny, sctrl, afwMath.MEANCLIP)
+    bkgd = afwMath.makeBackground(diff, bctrl)
+    diff -= bkgd.getImageF(afwMath.Interpolate.CUBIC_SPLINE, afwMath.REDUCE_INTERP_ORDER)
+    dim0 = diff[0: -n, : -n,  afwImage.LOCAL].clone()
+    dim0 -= afwMath.makeStatistics(dim0, afwMath.MEANCLIP, sctrl).getValue()
+    w, h = dim0.getDimensions()
+    xcorr = afwImage.ImageD(n + 1, n + 1)
+    for di in range(n + 1):
+        for dj in range(n + 1):
+            dim_ij = diff[di:di + w, dj: dj + h, afwImage.LOCAL].clone()
+            dim_ij -= afwMath.makeStatistics(dim_ij, afwMath.MEANCLIP, sctrl).getValue()
+
+            dim_ij *= dim0
+            xcorr[di, dj] = afwMath.makeStatistics(dim_ij, afwMath.MEANCLIP, sctrl).getValue()
+    L = np.shape(xcorr.getArray())[0]-1
+    XCORR = np.zeros([2*L+1, 2*L+1])
+    for i in range(L+1):
+        for j in range(L+1):
+            XCORR[i+L, j+L] = xcorr.getArray()[i, j]
+            XCORR[-i+L, j+L] = xcorr.getArray()[i, j]
+            XCORR[i+L, -j+L] = xcorr.getArray()[i, j]
+            XCORR[-i+L, -j+L] = xcorr.getArray()[i, j]
+    return xcorr, np.sum(means1)
+
+
+def simulateXcorr(fluxLevels, imageShape, addCorrelations=False, correlationStrength=0.1, nSigma=5, border=3,
```

Comment:
I don't see any use of this in the package. Can you get rid of it, along with _crossCorrelateSimulate above? If you have unresolved issues to work out then at least file a ticket and add a TODO message to the doc string for both of these functions and consider moving it to a separate module.

Reply:
OK, this has been tidied up significantly. The code must live on, and I think this is the right place for it, but it certainly needs a little more work, as there are some known problems. I've written a new function that unifies the generating images part and allows switching which code is called to do the calculations, and filed DM-15756 to finish up that bit of work, which will then remove the vestigial function here I think. If you have any comments on how you think that should go (including if you disagree that it should be in this module) do please leave them on the ticket, but I hope this is OK for now.

Reply:
That sounds fine to me.